src/checkers/inference/solver/strategy/GraphSolvingStrategy.java

```diff
@@ -124,77 +124,95 @@ protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Const
-        return solutions;
+        return results;
     }
 
     /**
      * This method is called if user wants to call all underlying solvers in sequence.
-     * 
+     *
      * @param underlyingSolvers
      * @return A list of Map that contains solutions from all underlying solvers.
      */
-    protected List<Map<Integer, AnnotationMirror>> solveInSequential(List<Solver<?>> underlyingSolvers) {
+    protected List<Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>>> solveInSequential(List<Solver<?>> underlyingSolvers) {
 
-        List<Map<Integer, AnnotationMirror>> solutions = new ArrayList<>();
+        List<Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>>> results = new ArrayList<>();
 
         long solvingStart = System.currentTimeMillis();
         for (final Solver<?> underlyingSolver : underlyingSolvers) {
-            solutions.add(underlyingSolver.solve());
+            Map<Integer, AnnotationMirror> solution = underlyingSolver.solve();
+            if (solution != null) {
+                results.add(new Pair<>(solution, new HashSet<>()));
+            } else {
+                results.add(new Pair<>(solution, underlyingSolver.explainUnsatisfiable()));
+            }
         }
         long solvingEnd = System.currentTimeMillis();
         //TODO: Refactor way of recording statistic.
         StatisticRecorder.record(StatisticKey.OVERALL_SEQUENTIAL_SOLVING_TIME, (solvingEnd - solvingStart));
-        return solutions;
+        return results;
     }
 
     /**
      * This method merges all solutions from all underlying solvers.
-     * 
-     * @param inferenceSolutionMaps
-     * @return an InferenceSolution for the given slots/constraints
+     *
+     * @param inferenceResults
+     * @return an InferenceResult for the given slots/constraints
      */
-    protected InferenceSolution mergeSolution(List<Map<Integer, AnnotationMirror>> inferenceSolutionMaps) {
+    protected InferenceResult mergeInferenceResults(List<Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>>> inferenceResults) {
 
-        Map<Integer, AnnotationMirror> result = new HashMap<>();
+        Map<Integer, AnnotationMirror> solutions = new HashMap<>();
+        Set<Constraint> explanations = new HashSet<>();
 
-        for (Map<Integer, AnnotationMirror> inferenceSolutionMap : inferenceSolutionMaps) {
-            result.putAll(inferenceSolutionMap);
+        for (Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>> inferenceResult : inferenceResults) {
+            if (inferenceResult.fst != null) {
+                solutions.putAll(inferenceResult.fst);
+            } else {
+                // If any one solution is null, whole solutions should be null.
+                solutions = null;
+                explanations.addAll(inferenceResult.snd);
+                break;
+            }
+        }
+        StatisticRecorder.record(StatisticKey.ANNOTATOIN_SIZE, (long) solutions.size());
```

Comment:
@topnessman  Mier, here might through a NPE, as line 206 set `solutions` to `null`.

Why not directly return the result with `explanations` in the else branch on line 204?